File: src/interval.cpp
#include "interval.hpp"
#include <algorithm>

Interval::Interval(int s, int e): start(s), end(e) {}

int Interval::get_length() const{
    return std::max(0,end-start);
}

bool Interval::operator<(const Interval& other) const{
    return start < other.start;
}

bool Interval::overlaps(const Interval& other, bool include_ends) const{
    if(include_ends)
        return (start <= other.end) and (end >= other.start);
    return (start < other.end) and (end > other.start);
}

Interval Interval::merge(const Interval& other) const{
    return Interval(
        std::min(start,other.start),
        std::max(end,other.end)
    );
}


Interval Interval::intersect(const Interval& other) const{
    return Interval(
        std::max(start,other.start),
        std::min(end,other.end)
    );
}

bool Interval::operator==(const Interval& other) const{
    return (start == other.start) and (end == other.end);
}

bool Interval::operator!=(const Interval& other) const{
    return (start != other.start) or (end != other.end);
}

File: pyboo/interval.py
from typing import List


class Interval:
    """
    start < end
    """
    def __init__(self,start=None,end=None):
        self.start = start if start else 0
        self.end = end if end else 0

    @property
    def length(self):
        return max(0,self.end-self.start)

    def __lt__(self,other):
        return self.start < other.start

    def overlaps(self,other,include_ends=True) -> bool:
        if include_ends:
            return (self.start <= other.end) and (self.end >= other.start)
        return (self.start < other.end) and (self.end > other.start)

    def merge(self,other):
        """
        Merge intervals assuming they overlap
        """

        return Interval(min(self.start,other.start),
            max(self.end,other.end))

    def intersect(self,other):
        return Interval(max(self.start,other.start),
            min(self.end,other.end))

    def __eq__(self,other):
        return (self.start == other.start) and (self.end == other.end)

    def __str__(self):
        return "[{},{}]".format(self.start,self.end)